source/common/http/http1/codec_impl.cc

```diff
@@ -187,7 +188,28 @@ void StreamEncoderImpl::encodeData(Buffer::Instance& data, bool end_stream) {
   }
 }
 
-void StreamEncoderImpl::encodeTrailers(const HeaderMap&) { endEncode(); }
+void StreamEncoderImpl::encodeTrailers(const HeaderMap& trailers) {
+  // Trailers only matter if it is a chunk transfer encoding
+  // https://tools.ietf.org/html/rfc7230#section-4.4
+  if (chunk_encoding_) {
```

Comment:
Also I'm sorry if I missed this (I'm just doing a really quick scan of this PR) but for HTTP/1 aren't trailers disallowed by default, and only allowed if the request headers had TE: trailers?  Maybe that's the knob I'm used to seeing.
Ideally there would also be a Trailer header saying which keys would be in trailers but AFIK that one is only a SHOULD.

@PiotrSikora for spec correctness

Reply:
As far as I know, trailers are always allowed.

However, since the use of trailers is rather limited, and most clients didn't support them until very recently, some proxies and servers only emit them when `TE: trailers` is present in the request headers.

Reply:
gRPC server only emits trailers when TE: trailers is set.

Reply:
@wlhee there's some cases that Envoy adds `TE: trailers` such as grpc-web filter. In that case the upstream gRPC server always see `TE: trailers` so it will emit trailer while it is possible that the downstream client doesn't support it.

Reply:
@PiotrSikora 

That's a SHOULD in [RFC 7230 4.1.2](https://tools.ietf.org/html/rfc7230#section-4.1.2)

> Unless the request includes a TE header field indicating "trailers"
   is acceptable, as described in Section 4.3, a server SHOULD NOT
   generate trailer fields that it believes are necessary for the user
   agent to receive.

Reply:
@lizan the keyword here is "necessary", i.e. server should not generate trailer fields that would render response incomplete if removed by the intermediaries. The `TE: trailers` is used to probe end-to-end support for trailers, so when it's not available, servers have no way to know if the client will ever receive them.

The reasoning for it is well explained in the linked paragraph:

>   Unless the request includes a TE header field indicating "trailers"
>   is acceptable, as described in Section 4.3, a server SHOULD NOT
>   generate trailer fields that it believes are necessary for the user
>   agent to receive.  Without a TE containing "trailers", the server
>   ought to assume that the trailer fields might be silently discarded
>   along the path to the user agent.  This requirement allows
>   intermediaries to forward a de-chunked message to an HTTP/1.0
>   recipient without buffering the entire response.

In any case, that paragraph was removed in the upcoming revision of the HTTP spec:
https://httpwg.org/http-core/draft-ietf-httpbis-messaging-latest.html#rfc.section.7.1.2